### DISCUSSION TITLE
Is it just me, or do userscripts load slower after the update?

### DISCUSSION THREAD

**Discussion by gaveitatry2** [violentmonkey/violentmonkey#1184] [General]
I was using 2.12.4 and I updated to 2.12.9 around three to five days ago.  I immediately noticed a difference on Yahoo! Mail.  I created a user script for Yahoo! Mail to change the layout and appearance some.  And since the update, the userscript doesn't load fast enough so I always see the default layout.  I have to refresh the page to see the desired changes.  It doesn't always do that.  But a lot of the time it does.  Enough to make me think that 2.12.9 loads userscripts slower than 2.12.4 does.  I'm not expecting or asking you to improve this.  Just please be aware and maybe you can figure a way to make it faster.  Thank you.  I guess if it annoys me enough, I can always revert back to 2.12.4.

**Comment by gaveitatry2** [comment]
@tophf 

I'm only going to do this once.  I didn't want to share my userscript and I don't like showing people my emails and other personal information.  If you don't figure out the problem, I will just continue using 2.12.4 until it is broken.  And hopefully I won't be forced to switch to a different userscript add-on, because I really like Violentmonkey a lot.

I created two gifs.  One to show what my Yahoo Mail userscript does when I use 2.12.4.  And one to show what my Yahoo Mail userscript does when I use 2.12.8.  The reason I made the two gifs is because my Yahoo Mail userscript is specifically made for my personal computer and my personal settings.  If you were to try and use the Yahoo Mail userscript, you would see very different results.  That's because I am using a 1600x900 computer resolution and my Firefox Portable browser is always 1100x870.  Using any other window size would destroy the userscript.  Again, this userscript was never meant to be publicized.  It's just for my personal use.

Btw, I had reverted to 2.12.4 a long time ago, like as soon as 2.12.8 was available and I found that it would not load my userscript until I had reloaded the page (I want it to load without having to refresh the page).  I'm just now trying 2.12.8 again because I wanted to be using the most up-to-date version, but I'm still having this difficulty so I had to revert back to 2.12.4 again.

So here are the two gifs...

This shows how I want my Yahoo Mail to look.  This is what my userscript does in version 2.12.4:
https://imgur.com/eNKb1lA

This shows how the userscript does not load in 2.12.8 (until I refresh the page):
https://imgur.com/njdKMp9

And this is my actual userscript:
https://pastebin.com/vtLvRLrD

**Comment by tophf** [comment]
2.12.8 is an old version, it had a few bugs that we've fixed in the [latest beta release](https://github.com/violentmonkey/violentmonkey/releases) so try it please.

Also, waitForKeyElements is indeed not the best tool as it retries every 300ms thus potentially creating an ugly pause. A much faster solution would be to use MutationObserver. Note that those nodes that you modify directly (without `parent`) don't need waitForKeyElements/MutationObserver because you can do it via GM_addStyle:

```js
GM_addStyle(`
  [data-test-id=mail-left-rail] {
    min-width: 275px !important;
  }
  [data-test-id=mail-app-component] {
    min-width: 93px !important;
  }
  [data-test-id=folder-list] > ul > li.H_6MWH,
  [data-test-id=toggle-folder-tree], /* comment */
  [data-test-id=right-rail-hidead-btn],
  [data-test-id=folder-list] > div:nth-child(2) {
    display: none !important;
  }
  [data-test-id=folder-item-container] > a > span > span {
    min-width: 22px !important;
  }
`);
```

**Comment by gaveitatry2** [comment]
Can you convert some of my code to MutationObserver, so I know how it is done?  I'm sure I have some parent and child elements so it's something I need to know.  I tried to figure it out before and was completely at loss.  I never learned coding.  I just did trial and error using stuff I found online until I found something that worked and accomplished what I wanted.

**Comment by gaveitatry2** [comment]
Sorry for jumping to conclusions.  I just wanted to return my test results and found that I could not reply and that the thread seemed to be locked.  You didn't give me any prior indication that you were the mean type, so it caught me off guard.  But I was used to that sort of treatment on other GitHubs and I also thought maybe another collaborator on your GitHub did this.

So anyway, I wanted to say that I've only been using ViolentMonkey BETA for a short time so nothing is purely 100% conclusive yet, but it appears that my userscript is working as before on 2.12.4, so that's a good thing.

1.) I'm still trying to learn this MutationObserver stuff.  Like how to convert this into MutationObserver:
```
waitForKeyElements ("[data-test-id=folder-list] > ul > li > div > a > span:nth-child(2) > div > [data-test-id=icon-btn-input_arrow_down]", MoveFoldersDownArrow);
function MoveFoldersDownArrow (jNode) {jNode.parent().css({"position" : "relative", "left" : "12px"});}
```
2.) And I wanted to ask you, since non-parent modes can use GM_AddStyle, which would be better?  Using GM_AddStyle for direct modes and MutationObserver for non-direct modes.  Or using MutationObserver for all modes, direct and non-direct?  Whichever causes the userscript to load the fastest.

**Reply by tophf** [reply]
The fastest is GM_addStyle because it's handled by the browser itself. Start by using it and remove from your code those selectors that don't have `parent()`. If the speed is acceptable then you can stop there, otherwise look for libraries/wrappers for MutationObserver that make it easier to use in userscripts.

**Comment by gaveitatry2** [comment]
Okay, I am getting that you can't or won't help me with MutationObserver.  That's not a problem.  It's just that you gave such a good example of GM_AddStyle, that I thought you'd be able to do it with MO too.  That's fine.  I just created two threads on Reddit, so maybe someone on there can help me.

One more question.  Does ViolentMonkey BETA update itself?  Like if you make an update next week and then another update the week after that, will my ViolentMonkey BETA install those updates automatically (I have it set to automatically update) or will I have to manually install the latest version of BETA?

**Reply by tophf** [reply]
Regarding the update thing, I don't know, never tested it myself.

**Reply by gera2ld** [reply]
It should update itself since v2.12.9.4 in Firefox. Earlier versions may not be able to update itself because the link to check for updates was changed recently due to cache problems.

The Chrome version should work all the time.

**Comment by gaveitatry2** [comment]
Wow, the GM_AddStyle is so much better!  It loads the code instantly, right away.  Thanks!

**Comment by gaveitatry2** [comment]
I think that I made some progress....

Someone on a Reddit subreddit helped me change:
```

waitForKeyElements ("[data-test-id=folder-list] > ul > li > div > a > span:nth-child(2) > div > [data-test-id=icon-btn-input_arrow_down]", MoveFoldersDownArrow);
function MoveFoldersDownArrow (jNode) {jNode.parent().css({"position" : "relative", "left" : "12px"});}
```

To this:

```
new MutationObserver(() => {
const keyElem = document.querySelector("[data-test-id=folder-list] > ul > li > div > a > span:nth-child(2) > div > [data-test-id=icon-btn-input_arrow_down]");
if (keyElem) {keyElem.style.position = "relative"; keyElem.style.left = "12px";}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

The problem is that their MutationObserver script only targets this node:
[data-test-id=folder-list] > ul > li > div > a > span:nth-child(2) > div > [data-test-id=icon-btn-input_arrow_down]

But I want it to target the parent of that node.

In my waitForKeyElements, I could do that by using .parent().css, but I don't know how to do that in MutationObserver.

Sorry for the update.  It just calms my nerves talking about it.

**Reply by gera2ld** [reply]
Try `if (keyElem) { $(keyElem).parent().css(...) }`

**Reply by gaveitatry2** [reply]
Thanks for the response.

This is the code that the Reddit person gave me to use with parent nodes:

```
new MutationObserver((mutations) => {
const keyElem = document.querySelector("[data-test-smartview-type=UNREAD]");
if (keyElem && keyElem.parentElement) {
keyElem.parentElement.style.backgroundColor = "red";}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

And it wasn't working.  But then I reposted my question on Reddit and someone found out why it isn't working.  Apparently having this breaks the MutationObserver code, so I removed it:
`// @run-at       document-start`

So I'm in the middle of setting up PlayStation Now with mouse and keyboard, but I'm pretty sure the code works now.  I just don't know how to use it with grandparents or greatgrandparents yet.  I'm going to try your code later because it looks easier to use.  If your code works, then changing parent to grandparent would be really easy because then I'd only have to add an extra parent().  So I will give it a try later.  Thank you!

**Comment by gaveitatry2** [comment]
Nope.  I thought the code he gave me worked, but it didn't.  I had left the old waitForKeyElements code by accident so that is why I thought it was working.

**Comment by gaveitatry2** [comment]
I hate this.  It's been around around a year since I wanted to try MutationObserver, but maybe I should give up on it.  I don't know how much faster it is.  Maybe it isn't fast at all.  Might be the exact same.  But I've tried to find working examples online.  I've posted on Reddit.  I've posted on StackOverflow.  But I just can't get anywhere.  I wish I could use GM_AddStyle with parent, grandparent, and greatgrandparent nodes.  That would be perfect.  It's super fast.

**Comment by gaveitatry2** [comment]
I think that I have some working MutationObserver code now.  I just needed to remove the run-at document-start line to get the MutationObserver to work.  It was also suggested that I add an observer.disconnect to make it stop observing.

```
// ==UserScript==
// @name         Yahoo! Mail (TEST)
// @namespace    idgas01
// @include      *mail.yahoo.com*
// @grant        GM_addStyle
// ==/UserScript==

new MutationObserver((mutations) => {
const keyElem = document.querySelector("[data-test-smartview-type=UNREAD]");
if (keyElem) {
keyElem.parentElement.style.backgroundColor = "red";
observer.disconnect();}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

And if I want to make it a grandparent, then I just add a .parentElement:

```
// ==UserScript==
// @name         Yahoo! Mail (TEST)
// @namespace    idgas01
// @include      *mail.yahoo.com*
// @grant        GM_addStyle
// ==/UserScript==

new MutationObserver((mutations) => {
const keyElem = document.querySelector("[data-test-smartview-type=UNREAD]");
if (keyElem) {
keyElem.parentElement.parentElement.style.backgroundColor = "red";
observer.disconnect();}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

I was curious about how to get it working with JQuery instead of JS, but the following code didn't work for me.  I will try to figure it out later.

```
// ==UserScript==
// @name         Yahoo! Mail (TEST)
// @namespace    idgas01
// @include      *mail.yahoo.com*
// @require      http://ajax.googleapis.com/ajax/libs/jquery/2.1.0/jquery.min.js
// @grant        GM_addStyle
// ==/UserScript==

new MutationObserver((mutations,observer) => {
const keyElem = document.querySelector("[data-test-smartview-type=UNREAD]");
if (keyElem) {
$(keyElem).parent().css({"background-color" : "red"});
observer.disconnect();}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

And I'm pretty sure I know how to do more than one MutationObserver function.  This is what I am going to try:

```
// ==UserScript==
// @name         Yahoo! Mail (TEST)
// @namespace    idgas01
// @include      *mail.yahoo.com*
// @grant        GM_addStyle
// ==/UserScript==

new MutationObserver((mutations) => {
const FirstElement = document.querySelector("[data-test-smartview-type=UNREAD]");
if (FirstElement) {
FirstElement.parentElement.style.backgroundColor = "red";}
const SecondElement = document.querySelector("[data-test-folder-container=Archive]");
if (SecondElement) {
SecondElement.parentElement.style.backgroundColor = "blue";
observer.disconnect();}
}).observe(document.body, {attributes:true, childList:true, subtree:true});
```

I wish that I had some idea how much better, less laggy, or faster this MutationObserver is supposed to be than waitForKeyElements.  But I'll just put my trust in that I should use the MutationObserver method from now on.

**Reply by gaveitatry2** [reply]
Hey showed me how to make it even better, so that it isn't doing multiple checks.

```
// ==UserScript==
// @name         Yahoo! Mail (TEST)
// @namespace    idgas01
// @include      *mail.yahoo.com*
// @grant        GM_addStyle
// ==/UserScript==

new MutationObserver((mutations) => {

const FirstElement = document.querySelector("[data-test-smartview-type=UNREAD]");
const SecondElement = document.querySelector("[data-test-folder-container=Archive]");

if (FirstElement && SecondElement) {
FirstElement.parentElement.style.backgroundColor = "red";
SecondElement.parentElement.style.backgroundColor = "blue";
observer.disconnect();
}

}).observe(document.body, {attributes:true, childList:true, subtree:true});
```
